File: ios/Application/ExtractionPlanner.swift
import Foundation

final class ExtractionPlanner {
  func planExtraction(
    entries: [EntryDescriptor],
    include: [String]?,
    exclude: [String]?,
    limits: ExtractionLimits,
    destination: URL
  ) throws -> [ExtractionPlanItem] {
    guard entries.count <= limits.maxEntries else {
      throw ArchiveDomainError.extractionLimitExceeded(limit: "maxEntries", value: "\(entries.count)")
    }

    var planned: [ExtractionPlanItem] = []
    var totalBytes: UInt64 = 0
    var duplicateTracker = DuplicatePathPolicy()

    for entry in entries {
      let normalizedPath = try ArchivePath(raw: entry.path)
      _ = try duplicateTracker.check(normalizedPath.normalized)

      guard normalizedPath.components.count <= limits.maxPathDepth else {
        throw ArchiveDomainError.extractionLimitExceeded(
          limit: "maxPathDepth",
          value: "\(normalizedPath.components.count)"
        )
      }

      let pathBytes = normalizedPath.normalized.lengthOfBytes(using: .utf8)
      guard pathBytes <= limits.maxPathBytes else {
        throw ArchiveDomainError.extractionLimitExceeded(
          limit: "maxPathBytes",
          value: "\(pathBytes)"
        )
      }

      if let include = include {
        let matched = include.contains { pattern in
          fnmatch(pattern, normalizedPath.normalized, 0) == 0
        }
        guard matched else { continue }
      }

      if let exclude = exclude {
        let excluded = exclude.contains { pattern in
          fnmatch(pattern, normalizedPath.normalized, 0) == 0
        }
        guard !excluded else { continue }
      }

      guard entry.uncompressedSize <= limits.maxEntryUncompressedBytes else {
        throw ArchiveDomainError.extractionLimitExceeded(
          limit: "maxEntryUncompressedBytes",
          value: "\(entry.uncompressedSize)"
        )
      }

      if entry.compressedSize == 0 && entry.uncompressedSize > 0 {
        throw ArchiveDomainError.extractionLimitExceeded(
          limit: "maxCompressionRatio",
          value: "infinite"
        )
      }

      if entry.compressedSize > 0 && entry.uncompressedSize / entry.compressedSize > UInt64(limits.maxCompressionRatio) {
        throw ArchiveDomainError.extractionLimitExceeded(
          limit: "maxCompressionRatio",
          value: "\(entry.uncompressedSize / entry.compressedSize)"
        )
      }

      let (newTotalBytes, overflow) = totalBytes.addingReportingOverflow(entry.uncompressedSize)
      guard !overflow else {
        throw ArchiveDomainError.extractionLimitExceeded(
          limit: "maxTotalUncompressedBytes",
          value: "overflow"
        )
      }
      totalBytes = newTotalBytes
      guard totalBytes <= limits.maxTotalUncompressedBytes else {
        throw ArchiveDomainError.extractionLimitExceeded(
          limit: "maxTotalUncompressedBytes",
          value: "\(totalBytes)"
        )
      }

      planned.append(ExtractionPlanItem(entry: entry, outputPath: normalizedPath))
    }

    return planned
  }
}

struct ExtractionPlanItem {
  let entry: EntryDescriptor
  let outputPath: ArchivePath
}

private func fnmatch(_ pattern: String, _ string: String, _ flags: Int32) -> Int32 {
  var p = pattern
  var s = string

  while !p.isEmpty {
    let pChar = p.removeFirst()
    if pChar == "*" {
      if p.isEmpty { return 0 }
      while !s.isEmpty {
        if fnmatch(p, s, 0) == 0 { return 0 }
        s.removeFirst()
      }
      return -1
    } else if pChar == "?" {
      if s.isEmpty { return -1 }
      s.removeFirst()
    } else {
      if s.isEmpty || s.removeFirst() != pChar { return -1 }
    }
  }
  return s.isEmpty ? 0 : -1
}

File: android/src/main/java/com/mcodex/nitroarchive/application/ExtractionPlanner.kt
package com.mcodex.nitroarchive.application

import com.mcodex.nitroarchive.application.ports.EntryDescriptor
import com.mcodex.nitroarchive.domain.ArchiveDomainError
import com.mcodex.nitroarchive.domain.ArchivePath
import com.mcodex.nitroarchive.domain.DuplicatePathPolicy
import com.mcodex.nitroarchive.domain.ExtractionLimits

data class ExtractionPlan(
    val entries: List<EntryDescriptor>,
    val totalUncompressedBytes: Long,
    val totalEntries: Int
)

class ExtractionPlanner {

    fun plan(
        allEntries: List<EntryDescriptor>,
        include: List<String>?,
        exclude: List<String>?,
        limits: ExtractionLimits = ExtractionLimits.DEFAULT
    ): ExtractionPlan {
        val duplicatePaths = DuplicatePathPolicy()
        var filtered = allEntries.map { entry ->
            val path = try {
                ArchivePath.fromRaw(entry.path)
            } catch (e: IllegalArgumentException) {
                throw ArchiveDomainError.PathTraversal(entry.path)
            }
            if (path.depth > limits.maxPathDepth) {
                throw ArchiveDomainError.ExtractionLimitExceeded("maxPathDepth", "${path.depth}")
            }
            if (path.normalized.toByteArray(Charsets.UTF_8).size > limits.maxPathBytes) {
                throw ArchiveDomainError.ExtractionLimitExceeded("maxPathBytes", "${path.normalized.toByteArray(Charsets.UTF_8).size}")
            }
            duplicatePaths.check(path.normalized)
            entry.copy(path = path.normalized)
        }

        if (!include.isNullOrEmpty()) {
            filtered = filtered.filter { entry ->
                include.any { pattern -> matchesGlob(entry.path, pattern) }
            }
        }

        if (!exclude.isNullOrEmpty()) {
            filtered = filtered.filterNot { entry ->
                exclude.any { pattern -> matchesGlob(entry.path, pattern) }
            }
        }

        if (filtered.size > limits.maxEntries) {
            throw ArchiveDomainError.ExtractionLimitExceeded(
                limit = "maxEntries",
                value = "${filtered.size}"
            )
        }

        filtered.forEach { entry ->
            if (entry.uncompressedSize > limits.maxEntryUncompressedBytes) {
                throw ArchiveDomainError.ExtractionLimitExceeded(
                    limit = "maxEntryUncompressedBytes",
                    value = "${entry.uncompressedSize}"
                )
            }
            if (entry.compressedSize == 0L && entry.uncompressedSize > 0L) {
                throw ArchiveDomainError.ExtractionLimitExceeded(
                    limit = "maxCompressionRatio",
                    value = "infinite"
                )
            }
            if (entry.compressedSize > 0L && entry.uncompressedSize / entry.compressedSize > limits.maxCompressionRatio) {
                throw ArchiveDomainError.ExtractionLimitExceeded(
                    limit = "maxCompressionRatio",
                    value = "${entry.uncompressedSize / entry.compressedSize}"
                )
            }
        }

        val totalBytes = filtered.sumOf { it.uncompressedSize }
        if (totalBytes > limits.maxTotalUncompressedBytes) {
            throw com.mcodex.nitroarchive.domain.ArchiveDomainError.ExtractionLimitExceeded(
                limit = "maxTotalUncompressedBytes",
                value = "$totalBytes"
            )
        }

        return ExtractionPlan(
            entries = filtered,
            totalUncompressedBytes = totalBytes,
            totalEntries = filtered.size
        )
    }

    private fun matchesGlob(path: String, pattern: String): Boolean {
        val regex = pattern
            .replace(".", "\\.")
            .replace("**", ".+")
            .replace("*", "[^/]+")
            .replace("?", "[^/]")
            .toRegex()
        return regex.matches(path)
    }
}
